File: advent/2021/08_2.py
import fileinput

total = 0

for line in fileinput.input():
    l, r = line.rstrip().split(" | ")

    cue = [""] * 10
    lp = [set(x) for x in l.split(' ')]
    rp = [set(x) for x in r.split(' ')]

    for s in lp:
        ln = len(s)
        if ln == 2:
            cue[1] = s
        elif ln == 3:
            cue[7] = s
        elif ln == 4:
            cue[4] = s
        elif ln == 7:
            cue[8] = s
        else:
            pass
        
    for s in lp:
        if len(s) == 6:
            if len(s & cue[1]) == 1:
                cue[6] = s
            elif len(s & cue[4]) == 3:
                cue[0] = s
            else:
                cue[9] = s
    
    for s in lp:
        if len(s) == 5:
            if len(s & cue[1]) == 2:
                cue[3] = s
            elif len(s & cue[6]) == 5:
                cue[5] = s
            else:
                cue[2] = s
    
    num = 0
    for o in rp:
        num *= 10
        for i in range(1, 10):
            if o == cue[i]:
                num += i

    total += num

print(total)

File: advent/2021/08_2.rs
use std::collections::{HashMap, HashSet};
use std::io::{self, BufRead};

fn main() {
    let tot: u32 = io::stdin()
        .lock()
        .lines()
        .map(|line| {
            let line = line.unwrap();
            let mut sp = line.split(" | ");

            let pat = sp.next().unwrap();
            let mut pat = pat
                .split(" ")
                .map(|s| s.chars().collect::<HashSet<_>>())
                .collect::<Vec<_>>();

            let mut seg = HashMap::<u8, _>::new();
            seg.insert(
                1,
                pat.swap_remove(pat.iter().position(|h| h.len() == 2).unwrap()),
            );
            seg.insert(
                4,
                pat.swap_remove(pat.iter().position(|h| h.len() == 4).unwrap()),
            );
            seg.insert(
                7,
                pat.swap_remove(pat.iter().position(|h| h.len() == 3).unwrap()),
            );
            seg.insert(
                8,
                pat.swap_remove(pat.iter().position(|h| h.len() == 7).unwrap()),
            );

            let v: Vec<_> = pat
                .iter()
                .enumerate()
                .filter(|(_, h)| h.len() == 6)
                .map(|(i, h)| {
                    let n = if seg[&1].intersection(h).count() == 1 {
                        6
                    } else if seg[&4].intersection(h).count() == 3 {
                        0
                    } else {
                        9
                    };
                    (i, n)
                })
                .collect();
            for (i, n) in v.into_iter().rev() {
                seg.insert(n, pat.swap_remove(i));
            }

            let v: Vec<_> = pat
                .iter()
                .enumerate()
                .filter(|(_, h)| h.len() == 5)
                .map(|(i, h)| {
                    let n = if seg[&1].intersection(h).count() == 2 {
                        3
                    } else if seg[&6].intersection(h).count() == 5 {
                        5
                    } else {
                        2
                    };
                    (i, n)
                })
                .collect();
            for (i, n) in v.into_iter().rev() {
                seg.insert(n, pat.swap_remove(i));
            }

            let mut num = 0;
            for r in sp
                .next()
                .unwrap()
                .split(" ")
                .map(|s| s.chars().collect::<HashSet<char>>())
            {
                let (k, _) = seg.iter().find(|(_, v)| &r == *v).unwrap();
                num = num * 10 + *k as u32;
            }
            num
        })
        .sum();

    println!("{}", tot);
}
